File: Submissions/19_Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int depth = recursive(head, n);
        if(depth == n) return head->next;
        return head;
    }
    
    int recursive(ListNode* node, int n) {
        if(node == nullptr) return 0;
        int depth = recursive(node->next, n);
        if(n == depth) {
            node->next = node->next->next;
        }
        
        return depth + 1;
    }
};

File: Kotlin/19_Remove_Nth_Node_From_End_of_List.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        val depth = recursive(head, n)
        
        if(depth == n) return head?.next
        
        return head
    }
    
    fun recursive(node: ListNode?, n: Int): Int {
        if(node == null) return 0
        
        val depth = recursive(node.next, n)
        if(n == depth) node.next = node.next.next
        
        return depth + 1
    }
}